polygon/bor/bor.go

```diff
@@ -272,8 +272,19 @@ func ValidateHeaderTime(
 	config *borcfg.BorConfig,
 	signaturesCache *lru.ARCCache[common.Hash, common.Address],
 ) error {
-	if header.Time > uint64(now.Unix()) {
-		return fmt.Errorf("%w: expected: %s(%s), got: %s", consensus.ErrFutureBlock, time.Unix(now.Unix(), 0), now, time.Unix(int64(header.Time), 0))
+	if config.IsBhilai(header.Number.Uint64()) {
+		// Don't waste time checking blocks from the future but allow a buffer of block time for
+		// early block announcements. Note that this is a loose check and would allow early blocks
+		// from non-primary producer. Such blocks will be rejected later when we know the succession
+		// number of the signer in the current sprint.
+		if header.Time-config.CalculatePeriod(header.Number.Uint64()) > uint64(time.Now().Unix()) {
+			return fmt.Errorf("%w: expected: %s(%s), got: %s", consensus.ErrFutureBlock, time.Unix(now.Unix(), 0), now, time.Unix(int64(header.Time), 0))
```

Comment:
Just a suggestion -- errors with unix based timestamps are much easier to debug and understand what caused it instead of printing full time stamps. 
Reference: https://github.com/maticnetwork/bor/pull/1545/commits/f735312bf0ffd3856b0b30a1e4221fa531621ecf